src/connectivity/all_pairs_all_simple_paths.rs

```diff
@@ -0,0 +1,94 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use rayon::prelude::*;
+
+use petgraph::algo;
+use petgraph::graph::NodeIndex;
+use petgraph::EdgeType;
+
+use crate::iterators::{AllPairsMultiplePathMapping, MultiplePathMapping};
+use crate::StablePyGraph;
+use retworkx_core::dictmap::*;
+
+pub fn all_pairs_all_simple_paths<Ty: EdgeType + Sync>(
+    graph: &StablePyGraph<Ty>,
+    min_depth: Option<usize>,
+    cutoff: Option<usize>,
+) -> AllPairsMultiplePathMapping {
+    if graph.node_count() == 0 {
+        return AllPairsMultiplePathMapping {
+            paths: DictMap::new(),
+        };
+    } else if graph.edge_count() == 0 {
+        return AllPairsMultiplePathMapping {
+            paths: graph
+                .node_indices()
+                .map(|i| {
+                    (
+                        i.index(),
+                        MultiplePathMapping {
+                            paths: DictMap::new(),
+                        },
+                    )
+                })
+                .collect(),
+        };
+    }
+    let intermediate_min: usize = match min_depth {
+        Some(depth) => depth - 2,
+        None => 0,
+    };
+    let intermediate_cutoff = cutoff.map(|depth| depth - 2);
+    let node_indices: Vec<NodeIndex> = graph.node_indices().collect();
+    AllPairsMultiplePathMapping {
+        paths: node_indices
+            .par_iter()
+            .filter_map(|u| {
+                let out_paths = MultiplePathMapping {
+                    paths: node_indices
+                        .iter()
+                        .filter_map(|v| {
+                            let output: Vec<Vec<usize>> = algo::all_simple_paths(
+                                graph,
+                                *u,
+                                *v,
+                                intermediate_min,
+                                intermediate_cutoff,
+                            )
```

Comment:
I'm wondering if it's valuable to avoid this loop by modifying petgraph's `algo::all_simple_paths` implementation in order to directly support a list of targets and not just a single target, as NetworkX https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.simple_paths.all_simple_paths.html#networkx.algorithms.simple_paths.all_simple_paths does. This might save us from doing the same work many times.

Reply:
Yeah, that's a good idea. It would make it more efficient since we only do a single traversal from each source instead of `n` traversals. I'm not sure how much it'll speed things up in practice but it's worth doing while we're thinking about it. I'll add a custom version of `algo::all_simple_paths` to retworkx-core in a little while to do this.